### DISCUSSION TITLE
Checks taking literally hours; add a timeout?

### DISCUSSION THREAD

**Discussion by simoncozens** [fonttools/fontbakery#4301] [General]
Fonts are getting more and more complex; things like [Handjet](https://github.com/google/fonts/blob/main/ofl/handjet/Handjet%5BELGR%2CELSH%2Cwght%5D.ttf), for example, have thousands of points and hundreds of masters.

We have a number of checks which decompile and recompile fonts (such as `hinting_stats` and `varfont/generate_static`) which take a ridiculously long time on Handjet; another check, `interpolation_issues`, takes *11 hours* on this font (I counted).

The annoying thing is that these checks are not something we can optimise in Fontbakery; the underlying operations are all FontTools routines: `ttFont.save()`, `mutator.instantiateVariableFont()` and `varLib.interpolatable.test`.

Of course this is a pathological font, but we want to be able to deal with arbitrary fonts. It's no good if you throw a directory of fonts at fontbakery and come back six hours later to discover that your CI job has died and you got no output. I figure our options are:

1. Replace these particular checks with Rust equivalents, or
2. Add a timeout for each check.

I think (2) is possible but it would probably involve messing around with signal handlers and I don't know how well that would play with `-j` (and Windows). It's probably the lowest cost approach in the short term, but fonts will only get more complicated, and we will have to do (1) eventually.

**Comment by davelab6** [comment]
Can't we just add a timeout to fb itself and run it per family, rather than at a directory of many families, in such a case?  Or just have a warning when the list of fonts fed to the main process is larger than 18?

**Comment by behdad** [comment]
I can take a look and see if the `interpolatable` algorithms can be sped up. But yeah if someone's up to porting that to Rust that would be even better.

**Comment by behdad** [comment]
> I can take a look and see if the `interpolatable` algorithms can be sped up. But yeah if someone's up to porting that to Rust that would be even better.

I can fix that fairly easily. Need to shuffle some code around. Only one of the glyphs (`pixel`) has variations (over 100 masters). The other glyphs have only one master. No need to instantiate and test them all at all the over-100 master locations for the font.

**Comment by behdad** [comment]
Fix up in https://github.com/fonttools/fonttools/pull/3300

**Reply by simoncozens** [reply]
That fixes the interpolation errors check, but the other two are still very slow.

**Reply by behdad** [reply]
Can you tell how long the other two take? I'll take a look at speeding up saving if possible.

**Reply by simoncozens** [reply]
Saving the font after it's decompiled takes 25 seconds:

```
>>> start = timer() ; font.save("foo.ttf"); end = timer(); print(end-start)
24.827587366104126
```

Instantiating takes 48s

```
$ time fonttools varLib.instancer Handjet\[ELGR,ELSH,wght\].ttf wght=400
Restricting axes: {'wght': (400, 400, 400)}
Loading variable font
Restricted limits: {'wght': (400, 400, 400)}
Normalized limits: {'wght': (0, 0, 0, 300, 500)}
Instantiating glyf/gvar tables
Instantiating HVAR table
Instantiating STAT table
Instantiating fvar table
Pruning name table
Saving partial variable font Handjet[ELGR,ELSH,wght]-partial.ttf
fonttools varLib.instancer Handjet\[ELGR,ELSH,wght\].ttf wght=400  48.51s user 0.18s system 99% cpu 48.820 total
```

**Comment by behdad** [comment]
> Saving the font after it's decompiled takes 25 seconds:

PR for speeding up saving: https://github.com/fonttools/fonttools/pull/3301

For me, brings it down from 11s to 0.22s.

**Comment by behdad** [comment]
> Instantiating takes 48s

Same PR (https://github.com/fonttools/fonttools/pull/3301) brings down instancing speed for me from 23s down to 0.36s.

**Comment by yisibl** [comment]
Any plans to rewrite fontbakery using Rust?

**Reply by simoncozens** [reply]
Plans, no. Experiments, certainly. ;-)

**Comment by simoncozens** [comment]
This is all better now.
